demos/bluetooth/mesh/thingy52/thingy52_music_cli/src/model_handler.c

```diff
@@ -0,0 +1,355 @@
+
+	for (int i = GREEN_LED; i <= RED_LED; i++) {
+		err |= sx1509b_pwm_pin_configure(io_expander, i);
+	}
+
+	if (err) {
+		printk("Initializing buttons and leds failed\n");
+	}
+}
+
+static void pdm_event_handler(nrfx_pdm_evt_t const *p_evt)
+{
+	if (p_evt->error) {
+		printk("PDM overflow error\n");
+	} else if (p_evt->buffer_requested && p_evt->buffer_released == 0) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_a,
+				    sizeof(mic_data.pdm_buf_a) /
+					    sizeof(mic_data.pdm_buf_a[0]));
+	} else if (p_evt->buffer_requested &&
+		   p_evt->buffer_released == mic_data.pdm_buf_a) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_b,
+				    sizeof(mic_data.pdm_buf_b) /
+					    sizeof(mic_data.pdm_buf_b[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_a,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	} else if (p_evt->buffer_requested &&
+		   p_evt->buffer_released == mic_data.pdm_buf_b) {
+		nrfx_pdm_buffer_set(mic_data.pdm_buf_a,
+				    sizeof(mic_data.pdm_buf_a) /
+					    sizeof(mic_data.pdm_buf_a[0]));
+
+		if (!mic_data.fft_result_ready) {
+			arm_rfft_q15(&S15, (q15_t *)mic_data.pdm_buf_b,
+				     mic_data.fft_buf);
+			arm_cmplx_mag_q15(mic_data.fft_buf, mic_data.spectrum,
+					  SPECTRUM_SIZE);
+
+			mic_data.fft_result_ready = true;
+		}
+	}
+}
+
+static void send_color_helper(int iterator, uint16_t value)
+{
+	int err = 0;
+
+	static struct bt_mesh_thingy52_rgb_msg msg = {
+		.ttl = 0,
+		.delay = 0xFFFF,
+		.speaker_on = false,
+	};
+
+	/* Should match NUM_FREQ_BANDS */
+	if (iterator == 0) {
+		msg.color.red = 0;
+		msg.color.green = 0;
+		msg.color.blue = value;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else if (iterator == 1) {
+		msg.color.red = 0;
+		msg.color.green = value;
+		msg.color.blue = 0;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else if (iterator == 2) {
+		msg.color.red = value;
+		msg.color.green = 0;
+		msg.color.blue = 0;
+		err |= bt_mesh_thingy52_cli_rgb_set(&thingy52_cli[iterator],
+						    NULL, &msg);
+	} else {
+		printk("Unknown iterator\n");
+	}
+
+	if (err) {
+		printk("Error sending RGB set message\n");
+	}
+}
+
+/* Values in the spectrum array are used as raw values (not q15) */
+static void microphone_work_handler(struct k_work *work)
+{
+	static uint16_t freq_sum[NUM_FREQ_BANDS];
+	static uint16_t prev_freq_sum[NUM_FREQ_BANDS];
+	static uint16_t total_sum;
+	static uint16_t checker;
+	static uint8_t divider;
+	static uint16_t adj_freq_sum;
+
+	if (mic_data.fft_result_ready) {
+		int i;
+
+		for (i = 0; i < NUM_FREQ_BANDS; i++) {
+			freq_sum[i] = 0;
+		}
+
+		/* Should create the amount of for-loops
+		 * that macthes NUM_FREQ_BANDS
+		 */
+		for (i = 0; i <= 80; i++) {
+			if (mic_data.spectrum[i] > 100) {
```

Comment:
Is this threshold necessary? All the bands are of the same size, couldn't you just sum every spectrum sample in each band? You'll also know the absolute max value of each band, and could scale the output accordingly, so you don't get the clipping on line 229

Reply:
Made a comment for this